li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java

```diff
@@ -339,82 +327,152 @@ synchronized public void unsubscribe() {
+    commitSync(Duration.ofMillis(_defaultApiTimeoutMs));
   }
 
   @Override
   synchronized public void commitSync(Duration timeout) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    executeCallbackWithTimeout("commitSync", null,
+        (consumer, ignored, timeoutDuration) -> consumer.commitSync(timeoutDuration), timeout);
   }
 
   @Override
   public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    commitSync(offsets, Duration.ofMillis(_defaultApiTimeoutMs));
   }
 
   @Override
   synchronized public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets, Duration timeout) {
-    throw new UnsupportedOperationException("Not implemented yet");
+   executeCallbackWithTimeout("commitSync", getPartitionValueMapByCluster(offsets),
+        (consumer, perClusterOffsets, timeoutDuration) -> {
+      consumer.commitSync((Map<TopicPartition, OffsetAndMetadata>) perClusterOffsets, timeoutDuration);
+        }, timeout);
+  }
+
+  // A wrapper callback for async offset commits, which will call the user-provided callback only after all per-cluster
+  // consumers finish commit and call this callback. If multiple consumers encounter exceptions while commiting offsets,
+  // one of them will be passed to the user-provided callback.
+  private class WrapperOffsetCommitCallback implements OffsetCommitCallback {
+    private AtomicInteger _pendingCommitCount;
+    // An aggregate of all offsets passed to each callback
+    private Map<TopicPartition, OffsetAndMetadata> _allOffsets;
+    private volatile Exception _exception;
+    private OffsetCommitCallback _userCallback;
+
+    public WrapperOffsetCommitCallback(AtomicInteger pendingCommitCount,
+        Map<TopicPartition, OffsetAndMetadata> allOffsets, Exception exception, OffsetCommitCallback userCallback) {
+      _pendingCommitCount = pendingCommitCount;
+      _allOffsets = allOffsets;
+      _exception = exception;
+      _userCallback = userCallback;
+    }
+
+    @Override
+    public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception exception) {
+      synchronized (_allOffsets) {
+        _allOffsets.putAll(offsets);
+      }
+      if (exception != null) {
+        _exception = exception;
+      }
+      if (_pendingCommitCount.decrementAndGet() == 0) {
+        if (_userCallback == null) {
+          // Copies what DefaultOffsetCommitCallback does, which is used by upstream consumer when a null callback is
+          // passed.
+          if (_exception != null) {
+              LOG.error("Offset commit with offsets {} failed", _allOffsets, _exception);
+          }
+        } else {
+          _userCallback.onComplete(_allOffsets, _exception);
+        }
+      }
+    }
   }
 
   @Override
-  public void commitAsync() {
-    throw new UnsupportedOperationException("Not implemented yet");
+  synchronized public void commitAsync() {
```

Comment:
Should we remove the synchronized here? Since the internal commitAsync(null) would grab the synchronized lock.

Reply:
I think one can safely call a synchronized method from another synchronized method, but removed anyway.